### PR TITLE
Implement custom constructor functionality

### PR DESCRIPTION
**Pull request by justing-bq**
Resolves https://github.com/amzn/ion-object-mapper-dotnet/issues/12.

By default, C# objects are created using the `Activator.CreateInstance(Type)` method which calls the default constructor. If you want to supply a constructor yourself, you can markup a constructor with the `IonConstructor` Attribute. Additionally, the parameters to the constructor must be specified using the `IonPropertyName` Attribute again:

```c#
public class Wheel
{
    private string specification;
    
    [IonConstructor]
    public Wheel([IonPropertyName("specification")] string specification)
    {
         this.specification = specification;
    }
}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 1]
```diff
@@ -124,6 +120,118 @@ public override void Serialize(IIonWriter writer, object item)
...
+
+            // Deserialize Ion and organize deserialized results into three categories:
+            // 1. Values to be passed into the Ion constructor.
+            // 2. Properties to be set after construction.
+            // 3. Fields to be set after construction.
+            var constructorArgs = new object[parameters.Length];
+            var remainingProperties = new List<(PropertyInfo, object)>();
+            var remainingFields = new List<(FieldInfo, object)>();
+            IonType ionType;
+            while ((ionType = reader.MoveNext()) != IonType.None)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:190]: I dislike that there are two methods of setting properties/fields on an object based on how it's created. I know this is done to avoid re-setting properties specified in the constructor, but can you refactor this and the common property/field specification to take in a set of names to ignore, similar to the precedence ordering?

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 2]
```diff
@@ -25,6 +25,19 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
 
         public override object Deserialize(IIonReader reader)
         {
+            var ionConstructors = targetType.GetConstructors(BINDINGS).Where(IsIonConstructor);
+            if (ionConstructors.Any())
+            {
+                if (ionConstructors.Count() > 1)
+                {
+                    throw new NotSupportedException(
+                        $"More than one constructor in class {targetType.Name} " +
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:34]: Can you reword this to be a bit more clear, something like "Only one constructor in class <blah> may be annotated with the ..."

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 3]
```diff
@@ -35,36 +48,19 @@ public override object Deserialize(IIonReader reader)
                 FieldInfo field;
                 if (property != null)
                 {
-                    if (IsReadOnlyProperty(property))
+                    var deserialized = new object();
+                    if (this.DeserializeProperty(property, reader, ionType, ref deserialized))
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:52]: When we return a boolean to indicate success, the method name should be prefixed with `Try`, so this would become `TryDeserializeProperty`.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 4]
```diff
@@ -124,6 +120,118 @@ public override void Serialize(IIonWriter writer, object item)
...
+            if (IsReadOnlyProperty(property))
+            {
+                // property.SetValue() does not work with a readonly property.
+                // logic for handling deserializing readonly properties happens during field processing
+                // when we detect backing fields for the property.
+                return false;
+            }
+
+            deserialized = ionSerializer.Deserialize(reader, property.PropertyType, ionType);
+            if (options.IgnoreDefaults && deserialized == default)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:134]: You can just return the inverted value here instead of `if (return false) else return true`

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 5]
```diff
@@ -124,6 +120,118 @@ public override void Serialize(IIonWriter writer, object item)
...
+            {
+                return false;
+            }
+
+            return true;
+        }
+        
+        private bool DeserializeField(FieldInfo field, IIonReader reader, IonType ionType, ref object deserialized)
+        {
+            deserialized = ionSerializer.Deserialize(reader, field.FieldType, ionType);
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:144]: Can you move this deserialization down to just before you do the default check?

Comment (kylepbit) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:144]: Actually, do you have to consume the value from the reader to consume subsequent values? If so, then this is in the right place, but I wonder about the above method at line 125 since it may ignore deserialization.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 6]
```diff
@@ -124,6 +120,118 @@ public override void Serialize(IIonWriter writer, object item)
...
+            
+            if (options.IgnoreReadOnlyFields && field.IsInitOnly)
+            {
+                return false;
+            }
+            
+            if (options.IgnoreDefaults && deserialized == default)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:151]: This can return the inverted value of the `if` condition directly.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 7]
```diff
@@ -124,6 +120,118 @@ public override void Serialize(IIonWriter writer, object item)
...
+                }
+
+                paramIndexMap.Add(ionPropertyName.Name, i);
+            }
+
+            reader.StepIn();
+
+            // Deserialize Ion and organize deserialized results into three categories:
+            // 1. Values to be passed into the Ion constructor.
+            // 2. Properties to be set after construction.
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:184]: What about methods that are annotated?

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 8]
```diff
@@ -25,6 +25,19 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
 
         public override object Deserialize(IIonReader reader)
         {
+            var ionConstructors = targetType.GetConstructors(BINDINGS).Where(IsIonConstructor);
```
Comment (HomayounA) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:28]: You might want to add Take(2) at the end, for possible performance gain (in case there are more than two constructors).

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 2 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Done.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 3 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Done.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 4 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Done.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 6 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Addressed.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 5 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:140]: Moving the deserialization down resulted in a test failing with:
```
System.NotImplementedException: The method or operation is not implemented.
    at Amazon.IonDotnet.Internals.Text.TextScanner.SkipOverTimestamp()
```

Since this sorta thing is a possibility, I made deserialization happen at the beginning of both of these `TryDeserialize` methods.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 7 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:175]: That relies on implementation in another PR. I will add that logic once the other PR is merged in.

In the meantime I will downgrade this PR to a draft PR.

**Review by kylepbit — COMMENTED** [review]
+1

**Comment by justing-bq** [comment]
This PR still needs to add method deserialization logic to DeserializeWithIonConstructor() once #25 gets merged.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 9]
```diff
@@ -124,6 +120,118 @@ public override void Serialize(IIonWriter writer, object item)
             writer.StepOut();
         }
 
+        private bool DeserializeProperty(PropertyInfo property, IIonReader reader, IonType ionType, ref object deserialized)
+        {
```
Comment (HomayounA) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:124]: Can you add a comment header for this method to mention the intent.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 8 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Done.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 9 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:126]: Done.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 10]
```diff
@@ -124,6 +120,116 @@ public override void Serialize(IIonWriter writer, object item)
...
+            // We deserialize whether or not we ultimately use the result because values
+            // for some Ion types need to be consumed in order to advance the reader.
+            deserialized = ionSerializer.Deserialize(reader, property.PropertyType, ionType);
+            
+            if (IsReadOnlyProperty(property))
+            {
+                // property.SetValue() does not work with a readonly property.
+                // logic for handling deserializing readonly properties happens during field processing
+                // when we detect backing fields for the property.
+                return false;
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:136]: Reading this - will this then work? If you read the value but then continue if it's read only, then you cannot read it again when you're trying do deserialize the field.

Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:136]: In .NET, properties have backing fields under the hood. So when we encounter a readonly property, since we can't call `property.SetValue()`, we instead process that property's backing field (which does not have the same readonly constraints. ie. `field.SetValue()` works).

The backing field has a distinct name from the property (ie. a property named `Major` has a backing field called `<Major>k__BackingField`) and the field will have been serialized with that distinct name so we still get a chance to find the backing field during another iteration of the loop.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 11]
```diff
@@ -228,16 +238,19 @@ private object DeserializeWithIonConstructor(ConstructorInfo ionConstructor, IIo
 
             reader.StepIn();
 
-            // Deserialize Ion and organize deserialized results into three categories:
+            // Deserialize Ion and organize deserialized results into four categories:
             // 1. Values to be passed into the Ion constructor.
-            // 2. Properties to be set after construction.
-            // 3. Fields to be set after construction.
+            // 2. Values to be set via annotated setters after construction.
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:243]: Can you specifically mention set via methods?

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs [hunk 12]
```diff
@@ -172,13 +172,14 @@ public void ExceptionOnMultiParameterIonPropertySetterMethods()
         public void SerializesAndDeserializesObjectsWithIonConstructor()
         {
             var serializer = new IonSerializer();
-            var wheel = new Wheel("default", 17, "MSW", "black");
+            var wheel = new Wheel("default", 17, 20, "MSW", "black");
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:175]: Should this just use Check?

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 11 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Done.

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs (hunk 12 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs]: `Check` does not work here as the deserialized `Specification` property differs from the original (which is how we verify that the custom constructor was used).

I tried altering the Specification to be a field (which is ignored by default) and then let the custom constructor set the field but that does not work because the field is ignored during serialization and then absent entirely during deserialization (when the custom construction happens).

Comment (kylepbit) [reply Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs]: Can you verify that they are not equal then, if that is expected?

Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs]: We check the deserialized specification more precisely with 
`Assert.AreEqual($"Specification: {wheel.Specification}, Size: {wheel.Size} inches", deserialized.Specification);` 
(which also entails checking that they are not equal).

The remaining checks for `offset`, `Brand`, and `color` are to verify that the respective method, property, and field logic in `DeserializeWithIonConstructor()` behaves as expected.

File: Amazon.IonObjectMapper.Test/IonObjectSerializerTest.cs [hunk 13]
```diff
@@ -168,6 +168,29 @@ public void ExceptionOnMultiParameterIonPropertySetterMethods()
...
+            var stream = defaultSerializer.Serialize(wheel);
+            var deserialized = defaultSerializer.Deserialize<Wheel>(stream);
+
+            // Verify custom IonConstructor was used and remaining properties/fields were set after construction.
+            Assert.AreEqual(
+                $"Specification: {wheel.Specification}, Size: {wheel.Size} inches", deserialized.Specification);
+            Assert.AreEqual(wheel.offset, deserialized.offset);
+            Assert.AreEqual(wheel.Brand, deserialized.Brand);
+            Assert.AreEqual(wheel.color, deserialized.color);
+        }
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.IonObjectMapper.Test/IonObjectSerializerTest.cs:185]: `Check` does not work here as the deserialized `Specification` property differs from the original (which is how we verify that the custom constructor was used).

I tried altering the `Specification` to be a field (which is ignored by default) and then let the custom constructor set the field but that does not work because the field is ignored during serialization and then absent entirely during deserialization (when the custom construction happens).

[the remark this answers is not in the document]
Comment (alpian) [reply Amazon.IonObjectMapper.Test/IonObjectSerializerTest.cs:185]: You don't need to "prove the constructor was called" by having code in the constructor. You just need to have an object that can only be created via its constructor, and which set's private fields, and then . Then you can use `Check` as Homayoun suggestst.

File: Amazon.IonObjectMapper.Test/TestObjects.cs [hunk 14]
```diff
@@ -292,10 +292,64 @@ public override string ToString()
 
     public class Wheel
     {
+        public string Specification { get; init; }
+        public int Size { get; init; }
+        public string Brand { get; init; }
+
+        [IonField]
```
Comment (alpian) [inline Amazon.IonObjectMapper.Test/TestObjects.cs:299]: It will be easier if you just create brand new test objects. I don't see the need for an `IonField` here.

File: Amazon.IonObjectMapper.Test/TestObjects.cs [hunk 15]
```diff
@@ -292,10 +292,64 @@ public override string ToString()
...
+            this.offset = offset;
+        }
+    }
 
+    public class Tire
+    {
+        public string specification { get; init; }
+        public int size { get; init; }
+
+        [IonConstructor]
```
Comment (alpian) [inline Amazon.IonObjectMapper.Test/TestObjects.cs:339]: How does this work? There are no annotated params

File: Amazon.IonObjectMapper.Test/TestObjects.cs [hunk 16]
```diff
@@ -292,10 +292,64 @@ public override string ToString()
...
+        public Tire(string specification, int size)
+        {
+            this.specification = specification;
+            this.size = size;
+        }
+
+        [IonConstructor]
+        private Tire(
```
Comment (alpian) [inline Amazon.IonObjectMapper.Test/TestObjects.cs:347]: Why two constructors?

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 17]
```diff
@@ -25,6 +25,19 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
 
         public override object Deserialize(IIonReader reader)
         {
+            var ionConstructors = targetType.GetConstructors(BINDINGS).Where(IsIonConstructor).Take(2);
```
Comment (alpian) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:28]: Why do we only `Take(2)`

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 18]
```diff
@@ -25,6 +25,19 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
...
+            if (ionConstructors.Any())
+            {
+                if (ionConstructors.Count() > 1)
+                {
+                    throw new InvalidOperationException(
+                        $"Only one constructor in class {targetType.Name} may be annotated " +
+                        "with the [IonConstructor] attribute and more than one was detected");
+                }
+
+                return this.DeserializeWithIonConstructor(ionConstructors.First(), reader);
```
Comment (alpian) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:38]: The only thing that is special about `IonConstructor` is how we create an instance of the object. Everything else is the same. Therefore it's just line 47 that you need to do this special constructor thing on.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 19]
```diff
@@ -175,6 +164,150 @@ public override void Serialize(IIonWriter writer, object item)
...
+                    throw new NotSupportedException(
+                        $"Parameter '{parameters[i].Name}' is not specified with the [IonPropertyName] attribute " +
+                        $"for {targetType.Name}'s IonConstructor. All constructor arguments must be annotated " +
+                        "so we know which parameters to set at construction time.");
+                }
+
+                paramIndexMap.Add(ionPropertyName.Name, i);
+            }
+
+            reader.StepIn();
```
Comment (alpian) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:239]: The below looks like it is more or less duplicated from the root `Deserialize` method.

**Review by alpian — CHANGES_REQUESTED** [review]

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 20]
```diff
@@ -175,6 +164,150 @@ public override void Serialize(IIonWriter writer, object item)
...
+                MethodInfo method;
+                PropertyInfo property;
+                FieldInfo field;
+                if (paramIndexMap.ContainsKey(reader.CurrentFieldName))
+                {
+                    var index = paramIndexMap[reader.CurrentFieldName];
+                    var deserialized = ionSerializer.Deserialize(reader, parameters[index].ParameterType, ionType);
+                    constructorArgs[index] = deserialized;
+                }
+                else if ((method = FindSetter(reader.CurrentFieldName)) != null)
```
Comment (alpian) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:262]: Specifically, these are the duplicated parts, which we need to try to avoid.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 21]
```diff
@@ -175,6 +230,79 @@ public override void Serialize(IIonWriter writer, object item)
             writer.StepOut();
         }
 
+        // Deserialize the given method and return bool to indicate whether the deserialized result should be used.
+        private bool TryDeserializeMethod(MethodInfo method, IIonReader reader, IonType ionType, out object deserialized)
```
Comment (byronlin13) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:234]: This will always return true, unless the method is not a setter which we will throw

File: Amazon.IonObjectMapper.Test/TestObjects.cs [hunk 22]
```diff
@@ -292,10 +292,43 @@ public override string ToString()
 
     public class Wheel
     {
+        public string Brand { get; init; }
```
Comment (byronlin13) [inline Amazon.IonObjectMapper.Test/TestObjects.cs:295]: Can we create an object with an Ion constructor and with object parameters. I wonder if the current logic handles this.

**Comment by byronlin13** [comment]
Could you rebase so we can see the code coverage?

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 23]
```diff
@@ -25,69 +25,124 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
...
+                targetObject = options.ObjectFactory.Create(options, reader, targetType);
+            }
+
             reader.StepIn();
 
+            // Read the Ion and organize deserialized results into three categories:
+            // 1. Values to be set via annotated methods.
+            // 2. Properties to be set.
+            // 3. Fields to be set.
+            // Any of these deserialized results may also be used for the annotated constructor (if there is one).
```
Comment (byronlin13) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:63]: Is this comment accurate? The deserialized results are used to set the values, properties, fields of the object itself, not to the annotated constructor.

**Comment by justing-bq** [comment]
> Could you rebase so we can see the code coverage?

Rebased.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 23 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:63]: It's accurate. After the method/property/field logic we check if the current Ion field is one of the arguments for the annotated constructor and populate the `constructorArgs` array accordingly.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 21 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:234]: Correct. This was done for the sake of symmetry with the `TryDeserializeProperty` and `TryDeserializeField` methods which can return true or false.

File: Amazon.IonObjectMapper.Test/IonObjectSerializerTest.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.IonObjectMapper.Test/IonObjectSerializerTest.cs:175]: Now using `Check`.
(we are relying on the expectation that we would get a `No parameterless constructor defined for type 'Amazon.IonObjectMapper.Test.Wheel'` exception if this test invoked `Activator.CreateInstance` rather than the annotated constructor).

File: Amazon.IonObjectMapper.Test/TestObjects.cs (hunk 15 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper.Test/TestObjects.cs:319]: This `Tire` class is meant for testing the exception when we have multiple `[IonConstructor]`s so the unannotated params did not matter. Regardless I added the param annotations for the sake of completeness.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 17 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:35]: `Take(2)` was added for a slight performance enhancement as we don't need more than 2 to run the following logic (since 2 or more IonConstructors result in an exception).

File: Amazon.IonObjectMapper.Test/TestObjects.cs (hunk 14 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper.Test/TestObjects.cs]: The `Wheel` class was not previously used by anything else in this code until this PR.
And the class has now been simplified to just have two properties.

File: Amazon.IonObjectMapper.Test/TestObjects.cs (hunk 16 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper.Test/TestObjects.cs]: To test that we get an exception if multiple `[IonConstructor]`s are defined.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 18 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs]: The two deserialization implementations are now merged into a single implementation.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 19 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs]: Duplication removed.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 20 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs]: The two implementations have merged into one so there is no more duplication.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 24]
```diff
@@ -175,6 +230,79 @@ public override void Serialize(IIonWriter writer, object item)
...
+            
+            if (IsReadOnlyProperty(property))
+            {
+                // property.SetValue() does not work with a readonly property.
+                // logic for handling deserializing readonly properties happens during field processing
+                // when we detect backing fields for the property.
+                return false;
+            }
+
+            return !options.IgnoreDefaults || deserialized != default;
```
Comment (alpian) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:266]: `==` won't always work, I don't think. I think you need to do this: https://www.google.com/search?client=firefox-b-e&q=c%23+how+to+check+if+value+is+a+default. Since this just got moved, feel free to address that in a subsequent PR.

**Review by alpian — APPROVED** [review]
We can fix the things I mentioned in a subsequent PR.
